testsuite/tests/typing-unique/unique_mod_class.ml

```diff
@@ -0,0 +1,150 @@
+(* TEST
+   flags += "-extension unique"
+ * expect
+*)
+
+(* This file tests the interaction between a module/class and its surrounding environment *)
+
+let unique_id (unique_ x) = ignore x
+
+
+(* you cannot use env vars as unique in classes/objects  *)
+let texp_object () =
+  let x = "foo" in
+  object (self)
+  val bar = unique_ x
+  end;
+[%%expect{|
+val unique_id : unique_ 'a -> unit = <fun>
+Line 8, characters 20-21:
+8 |   val bar = unique_ x
+                        ^
+Error: Found a shared value where a unique value was expected
+  Hint: This identifier cannot be used uniquely,
+  because it is defined in a class.
+|}]
+
+(* you can use env vars as shared and many, but they might collide with the external uses *)
+let texp_object () =
+  let x = "foo" in
+  unique_id x;
+  object (self)
+  val bar = x
+  end;
+[%%expect{|
+Line 3, characters 12-13:
+3 |   unique_id x;
+                ^
+Error: This is used uniquely here so cannot be used twice.  Another use is
+Line 5, characters 12-13:
+5 |   val bar = x
+                ^
+
+|}]
+
+(* you are not allowed to use x uniquely inside the module *)
+let texp_letmodule () =
+  let x = "foo" in
+  let module Bar = struct
+    let y = unique_ x
+  end
+  in
+  ()
+[%%expect{|
+Line 4, characters 12-21:
+4 |     let y = unique_ x
+                ^^^^^^^^^
+Error: This value is used as unique but it is a value that is top-level
```

Comment:
This error message is confusing. I cannot decipher from this what the actual restriction is.

Reply:
How about:
> Error: This value is used as unique but it is a value that is outside of the current module or class.

Reply:
I like that. I'm not sure how to implement it though - do we construct a `exception Typecore.Error`? I would also add a constructor to `Typecore.error` for this new type of error. I'm fine with that - just want to make sure that this is ok.

Reply:
Yeah, it's ok. There are a few other examples of files raising error messages from other files.

Reply:
Ahh I just realized that there might be some circular dependency issues between `typecore.ml` and `uniqueness_analysis.ml` - I think it might just be easier to construct the message from scratch.